services/sse-api/src/sse_api/watcher.py

```diff
@@ -0,0 +1,208 @@
+    _watchers: Dict[str, HubCacheChangedEvent]
+
+    def _notify_change(
+        self,
+        *,
+        dataset: str,
+        operation: str,
+        hub_cache: Optional[DatasetHubCacheResponse],
+        suscriber: Optional[str] = None,
+    ) -> None:
+        hub_cache_value = HubCacheChangedEventValue(dataset=dataset, operation=operation, hub_cache=hub_cache)
+        for watcher, event in self._watchers.items():
+            if suscriber is None or suscriber == watcher:
+                event.set_value(hub_cache_value=hub_cache_value)
+
+    def _unsubscribe(self, uuid: str) -> None:
+        self._watchers.pop(uuid)
+
+    def _subscribe(self) -> Tuple[str, HubCacheChangedEvent]:
+        event = HubCacheChangedEvent()
+        uuid = uuid4().hex
+        self._watchers[uuid] = event
+        return (uuid, event)
+
+
+class HubCacheWatcher:
+    """
+    Utility to watch the value of the cache entries with kind 'dataset-hub-cache'.
+    """
+
+    _watch_task: asyncio.Task[None]  # <- not sure about the type
+
+    def __init__(self, client: AsyncIOMotorClient, db_name: str, collection_name: str) -> None:
+        self._client = client
+        self._collection = self._client[db_name][collection_name]
+        self._publisher = HubCachePublisher(_watchers={})
+
+    def run_initialization(self, suscriber: str) -> asyncio.Task[Any]:
+        return asyncio.create_task(self._init_loop(suscriber=suscriber))
+
+    def start_watching(self) -> None:
+        self._watch_task = asyncio.create_task(self._watch_loop())
+
+    async def stop_watching(self) -> None:
+        self._watch_task.cancel()
+        with contextlib.suppress(asyncio.CancelledError):
+            await self._watch_task
+
+    def subscribe(self) -> Tuple[str, HubCacheChangedEvent]:
+        """
+        Subscribe to random value changes for the given space.
+        The caller is responsible for calling `self.unsubscribe` to release resources.
+
+        Returns:
+            (str, RandomValueChangedEvent):
+                A 2-tuple containing a UUID and an instance of RandomValueChangedEvent.
+                RandomValueChangedEvent can be `await`ed to be notified of updates to the random value.
+                UUID must be passed when unsubscribing to release the associated resources.
+        """
+        return self._publisher._subscribe()
+
+    def unsubscribe(self, uuid: str) -> None:
+        """
+        Release resources allocated to subscribe to the random value changes.
+        """
+        pub = self._publisher
+        pub._unsubscribe(uuid)
+
+    async def _init_loop(self, suscriber: str) -> None:
+        """
+        publish an event for each initial dataset-hub-cache cache entry.
+
+        TODO: we don't want to send to all the suscribers
+        """
+
+        async for document in self._collection.find(
+            filter={"kind": HUB_CACHE_KIND},
+            projection={"dataset": 1, "content": 1, "http_status": 1},
```

Comment:
Did you validate that this kind of query has a matching index in the collection?

Reply:
I think so, because three indexes start with `kind`, which is the only field used in the filter, and I think mongo can always sort by `_id`, right?

https://github.com/huggingface/datasets-server/blob/049ec7af688dbbf7ec56c3d9a779a8a8680b013a/libs/libcommon/src/libcommon/simple_cache.py#L113-L118

Anyway, I'm interested in your analysis